File: Example/JPImageresizerView/JPHelper/UIAlertController+JPImageresizer.m
//
//  UIAlertController+JPImageresizer.m
//  JPImageresizerView_Example
//
//  Created by 周健平 on 2020/1/23.
//  Copyright © 2020 ZhouJianPing. All rights reserved.
//

#import "UIAlertController+JPImageresizer.h"

@implementation UIAlertController (JPImageresizer)

+ (void)showWithPreferredStyle:(UIAlertControllerStyle)preferredStyle
                         title:(NSString *)title
                       message:(NSString *)message
                       actions:(NSArray<UIAlertAction *> *)actions
            isNeedCancelAction:(BOOL)isNeedCancelAction {
    if (actions.count == 0) return;
    
    UIAlertController *alertCtr = [self alertControllerWithTitle:title message:message preferredStyle:preferredStyle];
    
    for (UIAlertAction *action in actions) {
        [alertCtr addAction:action];
    }
    
    if (isNeedCancelAction) {
        [alertCtr addAction:[UIAlertAction actionWithTitle:@"取消" style:UIAlertActionStyleCancel handler:nil]];
    }
    
    [JPKeyWindow.jp_topViewController presentViewController:alertCtr animated:YES completion:nil];
}

#pragma mark - Alert

+ (void)alertWithTitle:(NSString *)title
               message:(NSString *)message
               actions:(NSArray<UIAlertAction *> *)actions {
    [self alertWithTitle:title
                 message:message
                 actions:actions
      isNeedCancelAction:YES];
}

+ (void)alertWithTitle:(NSString *)title
               message:(NSString *)message
               actions:(NSArray<UIAlertAction *> *)actions
    isNeedCancelAction:(BOOL)isNeedCancelAction {
    [self showWithPreferredStyle:UIAlertControllerStyleAlert
                           title:title
                         message:message
                         actions:actions
              isNeedCancelAction:isNeedCancelAction];
}

#pragma mark - Sheet

+ (void)sheetWithActions:(NSArray<UIAlertAction *> *)actions {
    [self sheetWithTitle:nil
                 message:nil
                 actions:actions
      isNeedCancelAction:YES];
}

+ (void)sheetWithTitle:(NSString *)title
               message:(NSString *)message
               actions:(NSArray<UIAlertAction *> *)actions {
    [self sheetWithTitle:title
                 message:message
                 actions:actions
      isNeedCancelAction:YES];
}

+ (void)sheetWithTitle:(NSString *)title
               message:(NSString *)message
               actions:(NSArray<UIAlertAction *> *)actions
    isNeedCancelAction:(BOOL)isNeedCancelAction {
    [self showWithPreferredStyle:UIAlertControllerStyleActionSheet
                           title:title
                         message:message
                         actions:actions
              isNeedCancelAction:isNeedCancelAction];
}

+ (void)changeResizeWHScale:(void(^)(CGFloat resizeWHScale))handler
              isArbitrarily:(BOOL)isArbitrarily
              isRoundResize:(BOOL)isRoundResize {
    if (!handler) return;
    [self sheetWithActions:@[
        [UIAlertAction actionWithTitle:[NSString stringWithFormat:@"使用%@", isArbitrarily ? @"固定比例" : @"任意比例"] style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
            handler(-1);
        }],
        
        [UIAlertAction actionWithTitle:[NSString stringWithFormat:@"%@圆切", isRoundResize ? @"取消" : @""] style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
            handler(0);
        }],
        
        [UIAlertAction actionWithTitle:@"1 : 1" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
            handler(1);
        }],
        
        [UIAlertAction actionWithTitle:@"2 : 3" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
            handler(2.0 / 3.0);
        }],
        
        [UIAlertAction actionWithTitle:@"3 : 5" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
            handler(3.0 / 5.0);
        }],
        
        [UIAlertAction actionWithTitle:@"9 : 16" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
            handler(9.0 / 16.0);
        }],
        
        [UIAlertAction actionWithTitle:@"7 : 3" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
            handler(7.0 / 3.0);
        }],
        
        [UIAlertAction actionWithTitle:@"16 : 9" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
            handler(16.0 / 9.0);
        }],
    ]];
}

+ (void)changeBlurEffect:(void(^)(UIBlurEffect *blurEffect))handler {
    if (!handler) return;
    NSMutableArray<UIAlertAction *> *actions = [NSMutableArray array];
    
    [actions addObject:[UIAlertAction actionWithTitle:@"移除模糊效果" style:UIAlertActionStyleDestructive handler:^(UIAlertAction * _Nonnull action) {
        handler(nil);
    }]];
    
    [actions addObject:[UIAlertAction actionWithTitle:@"ExtraLight" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        handler([UIBlurEffect effectWithStyle:UIBlurEffectStyleExtraLight]);
    }]];
    
    [actions addObject:[UIAlertAction actionWithTitle:@"Light" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        handler([UIBlurEffect effectWithStyle:UIBlurEffectStyleLight]);
    }]];
    
    [actions addObject:[UIAlertAction actionWithTitle:@"Dark" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        handler([UIBlurEffect effectWithStyle:UIBlurEffectStyleDark]);
    }]];
    
    if (@available(iOS 10, *)) {
        [actions addObject:[UIAlertAction actionWithTitle:@"Regular" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
            handler([UIBlurEffect effectWithStyle:UIBlurEffectStyleRegular]);
        }]];
        
        [actions addObject:[UIAlertAction actionWithTitle:@"Prominent" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
            handler([UIBlurEffect effectWithStyle:UIBlurEffectStyleProminent]);
        }]];
        
        if (@available(iOS 13, *)) {
            [actions addObject:[UIAlertAction actionWithTitle:@"SystemUltraThinMaterial" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
                handler([UIBlurEffect effectWithStyle:UIBlurEffectStyleSystemUltraThinMaterial]);
            }]];
            
            [actions addObject:[UIAlertAction actionWithTitle:@"SystemThinMaterial" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
                handler([UIBlurEffect effectWithStyle:UIBlurEffectStyleSystemThinMaterial]);
            }]];
            
            [actions addObject:[UIAlertAction actionWithTitle:@"SystemMaterial" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
                handler([UIBlurEffect effectWithStyle:UIBlurEffectStyleSystemMaterial]);
            }]];
            
            [actions addObject:[UIAlertAction actionWithTitle:@"SystemThickMaterial" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
                handler([UIBlurEffect effectWithStyle:UIBlurEffectStyleSystemThickMaterial]);
            }]];
            
            [actions addObject:[UIAlertAction actionWithTitle:@"SystemChromeMaterial" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
                handler([UIBlurEffect effectWithStyle:UIBlurEffectStyleSystemChromeMaterial]);
            }]];
            
            [actions addObject:[UIAlertAction actionWithTitle:@"SystemUltraThinMaterialLight" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
                handler([UIBlurEffect effectWithStyle:UIBlurEffectStyleSystemUltraThinMaterialLight]);
            }]];
            
            [actions addObject:[UIAlertAction actionWithTitle:@"SystemThinMaterialLight" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
                handler([UIBlurEffect effectWithStyle:UIBlurEffectStyleSystemThinMaterialLight]);
            }]];
            
            [actions addObject:[UIAlertAction actionWithTitle:@"SystemMaterialLight" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
                handler([UIBlurEffect effectWithStyle:UIBlurEffectStyleSystemMaterialLight]);
            }]];
            
            [actions addObject:[UIAlertAction actionWithTitle:@"SystemThickMaterialLight" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
                handler([UIBlurEffect effectWithStyle:UIBlurEffectStyleSystemThickMaterialLight]);
            }]];
            
            [actions addObject:[UIAlertAction actionWithTitle:@"SystemChromeMaterialLight" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
                handler([UIBlurEffect effectWithStyle:UIBlurEffectStyleSystemChromeMaterialLight]);
            }]];
            
            [actions addObject:[UIAlertAction actionWithTitle:@"SystemUltraThinMaterialDark" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
                handler([UIBlurEffect effectWithStyle:UIBlurEffectStyleSystemUltraThinMaterialDark]);
            }]];
            
            [actions addObject:[UIAlertAction actionWithTitle:@"SystemThinMaterialDark" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
                handler([UIBlurEffect effectWithStyle:UIBlurEffectStyleSystemThinMaterialDark]);
            }]];
            
            [actions addObject:[UIAlertAction actionWithTitle:@"SystemMaterialDark" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
                handler([UIBlurEffect effectWithStyle:UIBlurEffectStyleSystemMaterialDark]);
            }]];
            
            [actions addObject:[UIAlertAction actionWithTitle:@"SystemThickMaterialDark" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
                handler([UIBlurEffect effectWithStyle:UIBlurEffectStyleSystemThickMaterialDark]);
            }]];
            
            [actions addObject:[UIAlertAction actionWithTitle:@"SystemChromeMaterialDark" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
                handler([UIBlurEffect effectWithStyle:UIBlurEffectStyleSystemChromeMaterialDark]);
            }]];
        }
    }
    
    [self sheetWithActions:actions];
}

+ (void)replaceObj:(void(^)(UIImage *image, NSData *imageData, NSURL *videoURL))handler {
    if (!handler) return;
    [self sheetWithActions:@[
        [UIAlertAction actionWithTitle:@"Girl" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
            NSInteger index = 1 + arc4random() % GirlCount;
            NSString *girlImageName = [NSString stringWithFormat:@"Girl%zd.jpg", index];
            handler([UIImage imageNamed:girlImageName], nil, nil);
        }],
        
        [UIAlertAction actionWithTitle:@"Kobe" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
            handler([UIImage imageNamed:@"Kobe.jpg"], nil, nil);
        }],
        
        [UIAlertAction actionWithTitle:@"Flowers" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
            handler([UIImage imageNamed:@"Flowers.jpg"], nil, nil);
        }],
        
        [UIAlertAction actionWithTitle:@"咬人猫舞蹈节选（视频）" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
            handler(nil, nil, [NSURL fileURLWithPath:JPMainBundleResourcePath(@"yaorenmao.mov", nil)]);
        }],
        
        [UIAlertAction actionWithTitle:@"Gem（GIF）" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
            handler(nil, [NSData dataWithContentsOfFile:JPMainBundleResourcePath(@"Gem.gif", nil)], nil);
        }],
        
        [UIAlertAction actionWithTitle:@"Dilraba（GIF）" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
            handler(nil, [NSData dataWithContentsOfFile:JPMainBundleResourcePath(@"Dilraba.gif", nil)], nil);
        }],
        
        [UIAlertAction actionWithTitle:@"系统相册" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
            [UIImagePickerController openAlbum:handler];
        }],
    ]];
}

+ (void)rotation:(void(^)(BOOL isClockwise))handler1
     toDirection:(void(^)(JPImageresizerRotationDirection direction))handler2{
    if (!handler1 || !handler2) return;
    [self sheetWithActions:@[
        [UIAlertAction actionWithTitle:@"顺时针旋转" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
            handler1(YES);
        }],
        
        [UIAlertAction actionWithTitle:@"逆时针旋转" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
            handler1(NO);
        }],
        
        [UIAlertAction actionWithTitle:@"垂直向上（0°）" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
            handler2(JPImageresizerVerticalUpDirection);
        }],
        
        [UIAlertAction actionWithTitle:@"水平向右（90°）" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
            handler2(JPImageresizerHorizontalRightDirection);
        }],
        
        [UIAlertAction actionWithTitle:@"垂直向下（180°）" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
            handler2(JPImageresizerVerticalDownDirection);
        }],
        
        [UIAlertAction actionWithTitle:@"水平向左（270°）" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
            handler2(JPImageresizerHorizontalLeftDirection);
        }],
    ]];
}

+ (void)changeMaskImage:(void(^)(UIImage *maskImage))handler1
      gotoMaskImageList:(void(^)(void))handler2
          isReplaceFace:(BOOL)isReplaceFace
   isCanRemoveMaskImage:(BOOL)isCanRemoveMaskImage {
    if (!handler1 || !handler2) return;
    
    NSMutableArray<UIAlertAction *> *actions = [NSMutableArray array];
    
    [actions addObject:[UIAlertAction actionWithTitle:@"蒙版素材列表" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        handler2();
    }]];
    
    [actions addObject:[UIAlertAction actionWithTitle:@"love" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        handler1([UIImage imageNamed:@"love.png"]);
    }]];
    
    [actions addObject:[UIAlertAction actionWithTitle:@"Supreme" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        handler1([UIImage imageNamed:@"supreme.png"]);
    }]];
    
    if (isReplaceFace) {
        [actions addObject:[UIAlertAction actionWithTitle:@"Face Mask" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
            handler1([UIImage imageNamed:@"DanielWuFace.png"]);
        }]];
    }
    
    if (isCanRemoveMaskImage) {
        [actions addObject:[UIAlertAction actionWithTitle:@"移除蒙版" style:UIAlertActionStyleDestructive handler:^(UIAlertAction * _Nonnull action) {
            handler1(nil);
        }]];
    }
    
    [self sheetWithActions:actions];
}
@end

File: Example/JPImageresizerView/Extension/UIAlertController+JPImageresizer.swift
//
//  UIAlertController+JPImageresizer.swift
//  JPImageresizerView_Example
//
//  Created by aa on 2023/7/31.
//  Copyright © 2023 ZhouJianPing. All rights reserved.
//

import UIKit
import JPBasic
import JPImageresizerView

@objc
extension UIAlertController {
    static func changeResizeWHScale(_ handler: @escaping (_ resizeWHScale: CGFloat) -> Void, isArbitrarily: Bool, isRoundResize: Bool, fromVC: UIViewController? = nil) {
        UIAlertController
            .build(.actionSheet)
            .addAction("使用\(isArbitrarily ? "固定比例" : "任意比例")") { handler(-1) }
            .addAction("\(isRoundResize ? "取消" : "")圆切") { handler(0) }
            .addAction("1 : 1") { handler(1) }
            .addAction("2 : 3") { handler(2.0 / 3.0) }
            .addAction("3 : 5") { handler(3.0 / 5.0) }
            .addAction("9 : 16") { handler(9.0 / 16.0) }
            .addAction("7 : 3") { handler(7.0 / 3.0) }
            .addAction("16 : 9") { handler(16.0 / 9.0) }
            .addCancel()
            .present(from: fromVC ?? rootVC)
    }
    
    static func changeEffect(_ handler: @escaping (_ effect: UIVisualEffect?) -> Void, fromVC: UIViewController? = nil) {
        let alertCtr = UIAlertController
            .build(.actionSheet)
            .addDestructive("移除模糊效果") { handler(nil) }
        if #available(iOS 26.0, *) {
            alertCtr
                .addAction("Glass Regular") { handler(UIGlassEffect(style: .regular)) }
                .addAction("Glass Clear") { handler(UIGlassEffect(style: .clear)) }
        }
        alertCtr
            .addAction("ExtraLight") { handler(UIBlurEffect(style: .extraLight)) }
            .addAction("Light") { handler(UIBlurEffect(style: .light)) }
            .addAction("Dark") { handler(UIBlurEffect(style: .dark)) }
            .addAction("Regular") { handler(UIBlurEffect(style: .regular)) }
            .addAction("Prominent") { handler(UIBlurEffect(style: .prominent)) }
            .addAction("SystemUltraThinMaterial") { handler(UIBlurEffect(style: .systemUltraThinMaterial)) }
            .addAction("SystemThinMaterial") { handler(UIBlurEffect(style: .systemThinMaterial)) }
            .addAction("SystemMaterial") { handler(UIBlurEffect(style: .systemMaterial)) }
            .addAction("SystemThickMaterial") { handler(UIBlurEffect(style: .systemThickMaterial)) }
            .addAction("SystemChromeMaterial") { handler(UIBlurEffect(style: .systemChromeMaterial)) }
            .addAction("SystemUltraThinMaterialLight") { handler(UIBlurEffect(style: .systemUltraThinMaterialLight)) }
            .addAction("SystemThinMaterialLight") { handler(UIBlurEffect(style: .systemThinMaterialLight)) }
            .addAction("SystemMaterialLight") { handler(UIBlurEffect(style: .systemMaterialLight)) }
            .addAction("SystemThickMaterialLight") { handler(UIBlurEffect(style: .systemThickMaterialLight)) }
            .addAction("SystemChromeMaterialLight") { handler(UIBlurEffect(style: .systemChromeMaterialLight)) }
            .addAction("SystemUltraThinMaterialDark") { handler(UIBlurEffect(style: .systemUltraThinMaterialDark)) }
            .addAction("SystemThinMaterialDark") { handler(UIBlurEffect(style: .systemThinMaterialDark)) }
            .addAction("SystemMaterialDark") { handler(UIBlurEffect(style: .systemMaterialDark)) }
            .addAction("SystemThickMaterialDark") { handler(UIBlurEffect(style: .systemThickMaterialDark)) }
            .addAction("SystemChromeMaterialDark") { handler(UIBlurEffect(style: .systemChromeMaterialDark)) }
            .addCancel()
            .present(from: fromVC ?? rootVC)
    }
    
    static func replaceObj(_ handler: @escaping (_ image: UIImage?, _ imageData: Data?, _ videoURL: URL?) -> Void, fromVC: UIViewController? = nil) {
        UIAlertController
            .build(.actionSheet)
            .addAction("Girl") { handler(UIImage.randomGirlImage, nil, nil) }
            .addAction("Kobe") { handler(UIImage.bundleImage("Kobe.jpg"), nil, nil) }
            .addAction("Flowers") { handler(UIImage.bundleImage("flowers.jpg"), nil, nil) }
            .addAction("咬人猫舞蹈节选（视频）") { handler(nil, nil, URL(fileURLWithPath: Bundle.main.path(forResource: "yaorenmao", ofType: "mov")!)) }
            .addAction("See the sea（视频）") { handler(nil, nil, URL(fileURLWithPath: Bundle.main.path(forResource: "seeTheSea", ofType: "mp4")!)) }
            .addAction("Gem（GIF）") { handler(nil, try! Data(contentsOf: URL(fileURLWithPath: Bundle.main.path(forResource: "Gem", ofType: "gif")!)), nil) }
            .addAction("Dilraba（GIF）") { handler(nil, try! Data(contentsOf: URL(fileURLWithPath: Bundle.main.path(forResource: "Dilraba", ofType: "gif")!)), nil) }
            .addAction("八爪丸（GIF）") {
                Task {
                    let gifImage = await UIImage.getCuteOctoGIFImage()
                    handler(gifImage, nil, nil)
                }
            }
            .addAction("系统相册") {
                ImagePicker.openAlbumForObject { result in
                    switch result {
                    case let .success(obj):
                        handler(nil, obj.imageData, obj.videoURL)
                    case .failure:
                        break
                    }
                }
            }
            .addCancel()
            .present(from: fromVC ?? rootVC)
    }
    
    static func rotation(_ handler1: @escaping (_ isClockwise: Bool) -> Void, toDirection handler2: @escaping (_ direction: JPImageresizerRotationDirection) -> Void, fromVC: UIViewController? = nil) {
        UIAlertController
            .build(.actionSheet)
            .addAction("顺时针旋转 ⤵️") { handler1(true) }
            .addAction("逆时针旋转 ⤴️") { handler1(false) }
            .addAction("垂直向上 ⬆️") { handler2(.verticalUpDirection) }
            .addAction("水平向右 ➡️") { handler2(.horizontalRightDirection) }
            .addAction("垂直向下 ⬇️") { handler2(.verticalDownDirection) }
            .addAction("水平向左 ⬅️") { handler2(.horizontalLeftDirection) }
            .addCancel()
            .present(from: fromVC ?? rootVC)
    }
    
    static func changeMaskImage(_ handler1: @escaping (_ maskImage: UIImage?) -> Void,
                                gotoMaskImageList handler2: @escaping () -> Void,
                                isCanRemoveMaskImage: Bool,
                                fromVC: UIViewController? = nil) {
        let alertCtr = UIAlertController.build(.actionSheet)
            .addAction("蒙版素材列表") { handler2() }
            .addAction("love") { handler1(UIImage.bundleImage("love.png")) }
            .addAction("Supreme") { handler1(UIImage.bundleImage("supreme.png")) }
            .addAction("Face Mask") { handler1(UIImage.bundleImage("DanielWuFace.png")) }
        if isCanRemoveMaskImage {
            alertCtr.addDestructive("移除蒙版") { handler1(nil) }
        }
        alertCtr
            .addCancel()
            .present(from: fromVC ?? rootVC)
    }
}
